### DISCUSSION TITLE
Zero-downtime minor upgrade of tools

### DISCUSSION THREAD

**Discussion by knu** [jdx/mise#3843] [Ideas]
Suppose you’ve run `mise use rust@latest python@3.11`, which is quite common.  You currently have rust@1.82.0 and python@3.11.10 installed, and you decide to upgrade them to their latest releases.  As soon as you start `mise upgrade`, you find that you can’t run rust or python until the new versions finish installing.  Your scheduled background jobs are failing.  This occurs because the upgrade process immediately creates installs/rust/1.83.0 and installs/python/3.11.11 before the build stage completes, and mise picks those half-installed or currently empty versions, effectively breaking your environment during the upgrade.

I ended up writing a script like this: https://gist.github.com/knu/86c3d1a2c4cbcf02b1d6306fdffd8ded

Could there be support for this situation on the mise side?

**Comment by JamBalaya56562** [comment]
The breakage you described does not happen any more, and I measured the part that matters for your scheduled jobs rather than just the version-selection logic.

A scheduled job reaches the tool through the runtime symlink that `mise activate`/shims put on PATH, so I sampled that directly — `installs/node/22/bin/node` and `installs/go/1/bin/go` — in a tight loop while a **two-tool** upgrade ran with a cold cache, so both downloads were real:

```console
=== multi-tool upgrade, sampling the runtime symlinks in a tight loop ===
samples: 93
samples where node/22/bin/node did not run: 0
samples where go/1/bin/go did not run:      0

=== distinct states seen ===
v22.14.0 | go1.24.0 | ./22.14.0 | ./1.24.0
v22.14.0 | go1.24.0 | ./22.23.2 | ./1.26.5     <- swap caught mid-sample
v22.23.2 | go1.26.5 | ./22.23.2 | ./1.26.5

Upgraded 2 tools:
  node 22.14.0 → 22.23.2
  go 1.24.0 → 1.26.5
```

The middle row is the interesting one: the binaries ran before the `readlink` in that same sample, so the symlinks flipped between the two — meaning the sampling did straddle the transition rather than stepping over it, and the tool was still usable on both sides of it.

The selection logic underneath is why. mise creates the new version directory together with an `incomplete` marker before fetching anything, and a version only counts as installed once that marker is gone. Constructing exactly that state confirms it:

```console
$ mkdir -p .../installs/node/22.99.0 && touch .../installs/node/22.99.0/incomplete
$ mise exec -- node --version
v22.23.2                      # not 22.99.0
$ mise ls node
node  22.23.2  .../mise.toml  22

$ rm .../installs/node/22.99.0/incomplete    # bare empty directory
$ mise exec -- node --version
v22.23.2
```

So neither the half-installed nor the empty-directory shape you ran into gets picked, and the runtime symlink stays on the old version until the new one is finished.

One limit worth stating: sampling cannot prove the absence of a window shorter than the gap between samples (roughly tens of milliseconds here). What it does show is that across an entire two-tool upgrade with real downloads, nothing was ever unusable.

Your gist should not be needed for this. If a slow source build like rust still shows a gap, that would be worth a fresh report — the mechanism above is meant to cover it, so a failure there would be a different bug.
